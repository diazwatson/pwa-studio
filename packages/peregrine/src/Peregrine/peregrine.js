import React from 'react';
import { arrayOf, oneOf } from 'prop-types';
import { ToastContextProvider } from '../Toasts';
import { WindowSizeContextProvider } from '../hooks/useWindowSize';

const contextProviders = {
    TOASTS: ToastContextProvider,
    WINDOW_SIZE: WindowSizeContextProvider
};

const Peregrine = ({ children, providers }) => {
    return providers.reduceRight((memo, providerKey) => {
        const ContextProvider = contextProviders[providerKey];
        return <ContextProvider>{memo}</ContextProvider>;
    }, children);
};

Peregrine.propTypes = {
    providers: arrayOf(oneOf(Object.keys(contextProviders)))
};

Peregrine.defaultProps = {
    providers: ['WINDOW_SIZE', 'TOASTS']
};

export default Peregrine;
