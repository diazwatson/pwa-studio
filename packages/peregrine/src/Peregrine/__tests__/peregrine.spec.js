import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import Peregrine from '../peregrine';

test('render wraps children with all peregrine context providers', () => {
    const shallowRenderer = new ShallowRenderer();
    const tree = shallowRenderer.render(
        <Peregrine>
            <i />
        </Peregrine>
    );

    expect(tree).toMatchSnapshot();
});

test('render wraps children with single peregrine context provider', () => {
    const shallowRenderer = new ShallowRenderer();
    const tree = shallowRenderer.render(
        <Peregrine providers={['TOASTS']}>
            <i />
        </Peregrine>
    );

    expect(tree).toMatchSnapshot();
});

test('fails to render with invalid context provider', () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(
        <Peregrine providers={['WAFFLES']}>
            <i />
        </Peregrine>
    );

    expect(console.error).toHaveBeenCalledWith(
        expect.stringContaining(
            'Invalid prop `providers[0]` of value `WAFFLES` supplied to `Peregrine`, expected one of ["TOASTS","WINDOW_SIZE"].'
        )
    );
});
