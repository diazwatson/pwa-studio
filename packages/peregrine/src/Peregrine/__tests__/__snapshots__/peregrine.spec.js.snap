// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render wraps children with all peregrine context providers 1`] = `
<WindowSizeContextProvider>
  <ToastContextProvider>
    <i />
  </ToastContextProvider>
</WindowSizeContextProvider>
`;

exports[`render wraps children with single peregrine context provider 1`] = `
<ToastContextProvider>
  <i />
</ToastContextProvider>
`;
